source/common/common/thread.cc

```diff
@@ -1,36 +1,99 @@
 #include "source/common/common/thread.h"
 
+#include <thread>
+
+#include "source/common/common/assert.h"
+#include "source/common/common/macros.h"
+
 namespace Envoy {
 namespace Thread {
 
-bool MainThread::isMainThread() {
-  // If threading is off, only main thread is running.
-  auto main_thread_singleton = MainThreadSingleton::getExisting();
-  if (main_thread_singleton == nullptr) {
-    return true;
+namespace {
+
+// Singleton structure capturing which thread is the main dispatcher thread, and
+// which is the test thread. This info is used for assertions around catching
+// exceptions and accessing data structures which are not mutex-protected, and
+// are expected only from the main thread.
+//
+// TODO(jmarantz): avoid the singleton and instead have this object owned
+// by the ThreadFactory. That will require plumbing the API::API into all
+// call-sites for isMainThread(), which might be a bit of work, but will make
+// tests more hermetic.
+struct ThreadIds {
+  // Determines whether we are currently running on the main-thread or
+  // test-thread. We need to allow for either one because we don't establish
+  // the full threading model in all unit tests.
+  bool inMainOrTestThread() const {
+    std::thread::id id = std::this_thread::get_id();
+    absl::MutexLock lock(&mutex_);
```

Comment:
Is this going to be called in every `ASSERT` around main thread? Will this force more serialization/contention in debug builds than we have today? I realize we don't care a ton about performance in this build, but we want concurrency behavior to be fairly regular. It might not make a huge difference if it's not called a ton, since the critical sections are short. Just a question for discussion really.

Reply:
I think it should be OK to use an atomic for the thread IDs and clear them when the use-count goes to zero., so that this isMainOrTestThread check can be done without a lock. Let me look into that.

Reply:
If the test thread and main thread instance is initialized before any worker thread has been spawned, can we assume that the thread id is read-only and won't cause any race?

Reply:
There can be races observing the atomic thread ID member vars, but all outcomes will be correct. If the comparison matches then there can be no race because the observing thread and the writing thread must be the same. If the comparison mismatches then it doesn't matter whether the testing thread sees the prior default value of the thread ID or the new one; either would mismatch.